### DISCUSSION TITLE
Separate section with exercises in each chapter

### DISCUSSION THREAD

**Discussion by ppolesiuk** [ppolesiuk/type-systems-notes#52] [Ideas]
Each week I spend some time to invent exercises for the course. Usually, I don't put all of them on the exercise list, because I respect students time. I propose to create in each chapter a separate section called "Exercises" where we can put those "published" and "unpublished" exercises. While it requires a little extra work to write exercises in two languages (the course in given in Polish), I see the following benefits.

- For me: it is better to keep all teaching materials in one place. It will be easier to give this course in next years.
- For students: these extra exercises may turn out to be interesting problems to think about. Moreover, I expect that problems on the final exam will be slight modifications of those exercises (published and unpublished).
- For academic community: notes and exercises may be helpful for others that teach/learn the theory of programming languages. Our notes popularize our point of view on the topic.

**Comment by forell** [comment]
On top of the reasons mentioned, the notes occasionally vaguely hint that something is "left as an exercise". They would be more complete if the actual, detailed exercise statements were part of the notes as well.

**Reply by ppolesiuk** [reply]
The notes would be even more complete if we put solutions to (some of) exercises in it. Do you think it is a good idea? Of course, it is impossible given the amount of work that we devote to these notes, but I'm just curious about your opinion. The main argument against, beside additional amount of work, is that well written solutions tempt to cheat during classes. On the other hand such solutions could cover some extra discussion that I usually try to conduct during classes (e.g., which approach is better?).

**Reply by forell** [reply]
I'm not sure. In the context of the Type Systems class, I believe that it's niche enough that people enrolled out of genuine interest, and would not be tempted to cheat. However, if we're serious about making this a viable resource for the broader community, this could be a deal breaker for a prospective instructor who, for whatever reason, doesn't feel like they can trust in their students' academic honesty quite as much. For example, you'll find that the [Preface to Software Foundations](https://softwarefoundations.cis.upenn.edu/lf-current/Preface.html#lab10) explicitly urges its readers to refrain from posting their solutions online, specifically citing its use as course material as the reason. On the other hand, including the solutions would make the notes better for self-learning. The discussion-style exercises you mentioned will suffer a lot without a classroom environment. Ultimately, this is a hard question that I don't have a good answer to.
